### DISCUSSION TITLE
Untitled

### DISCUSSION THREAD

**Discussion by gusbemacbe** [AppImage/AppImageKit#1322] [Q&A]
Without comment.

**Comment by probonopd** [comment]
I don't know much about Wayland except that it is not 100% compatible with X11, makes screen sharing complicated, and never works with full performance on my Nvidia based systems, so forgive me if this is a stupid question, but:

Why are there two different versions of the application needed? Wayland proponents tend to claim that X11 applications can run in XWayland.

Of course the author of the Espanso AppImage could make two separate AppImages, but that is not what I'd recommend.

If there are really two different versions of the application needed, then it would be best if the author of the Espanso AppImage would put both versions into the same AppImage, add an `if` statement in the `AppRun`, and run the Wayland version of the application in case Wayland is detected to be running on the system. As a result, users could just run the AppImage and it would execute the version needed for their particular window server without the user having to deal with multiple AppImages.

**Comment by BlueDrink9** [comment]
@gusbemacbe it's a little bit rude to tag people in unrelated issues. Just link to the project - it's clearer that way. Especially since autokey is neither mine nor elleria's

**Comment by probonopd** [comment]
> Espanso built for X11 will not work on Wayland due to xclip. Wayland asks for wl-copy.

I see. So ideally the Espanso AppImage would ship both and decide at runtime which one to use.

**Comment by TheAssassin** [comment]
I second building some unified AppImage that supports either system. With some scripting it should be possible to make some prototype.

**Comment by federico-terzi** [comment]
Hey everyone, thanks for the feedback! 

>  So ideally the Espanso AppImage would ship both and decide at runtime which one to use.

That's a really clever idea, I didn't realize it was possible. I'll investigate further, though IIRC, Espanso is not recommending AppImage on Wayland as it's not compatible with the required Linux capabilities (instead of asking the users for root access on Wayland, Espanso uses Linux capabilities to get the required permissions for a short initial period). As far as I remember from my tests, the way AppImage was structured didn't allow setting linux capabilities to the binary unless the user performed an extraction first

**Comment by probonopd** [comment]
Thanks @federico-terzi. The contents of an AppImage are read-only, as an AppImage is a read-only self-mounting disk image. If the application was installed to e.g., `/usr/bin` where normal users have no root rights to change anything (and I assume normal users can't change _Linux capabilities_ there either?), how would it work for Wayland there?

**Comment by TheAssassin** [comment]
(Short overview, not 100% correct, but you'll get the idea.)

Capabilities can be set on files to grant unprivileged processes (i.e., processes whose uid is nonzero which applies to all non-root processes) privileged access to some kernel APIs. This way, one can avoid hacks like `suid` and doesn't have to run an entire process as `root`. Since capabilities (just like `suid`) are stored in the filesystem metadata, they cannot be used when mounting squashfs images as a regular user. This would open all kinds of attack vectors and thus create a huge security problem. (AppImages' squashfs images are mounted with `nosuid` and a few similar flags for a reason.)

There are modern alternatives to capabilities, suid etc., e.g., `polkit`. I've never had to work with polkit, really, though, so I'm not sure these would work for this application. I'm relatively sure I've seen AppImages use polkit before, though. See [this FAQ](https://github.com/balena-io/etcher/blob/3ece1fd84112892fbb744553a6181da10a103209/docs/USER-DOCUMENTATION.md#no-polkit-authentication-agent-found-error-in-gnulinux) for instance.

@federico-terzi maybe you could elaborate a bit on why Wayland needs capabilities for regular applications? You seem to have some understanding of this. Or maybe just link to other resources? I'm curious.

**Comment by probonopd** [comment]
Uh, is this a Linuxism that Wayland now requires? How does this work on non-Linux POSIX systems?

**Comment by TheAssassin** [comment]
I don't think this may be 100% accurate. I've been running some GUI software packaged as an AppImage on a Wayland system today...

**Comment by probonopd** [comment]
...yes, it's not AppImage that requires this _Linux capabilities_ thingy ;-)

**Comment by probonopd** [comment]
You need to change your script so that it loads `espanso-x11` from the directory it is in, rather than from the user's current working directory (cwd).

In other words, your script seems to want to load `expanso-x11` from `./` rather than from `/tmp/.mount_Espans<xxxxxx>/usr/bin/`.

**Comment by ppvnf** [comment]
any update on this?

**Comment by TheAssassin** [comment]
@ppvnf your question makes no sense. Please specify.

That said, this is more of a question. Moving to discussions.
